libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,550 @@
+                                        uint16_t * pThingNameLength )
+{
+    uint16_t index = 0U;
+    ShadowStatus_t returnStatus = SHADOW_STATUS_THINGNAME_PARSE_FAILED;
+
+    for( ; index < stringLength; index++ )
+    {
+        if( pString[ index ] == ( char ) '/' )
+        {
+            break;
+        }
+    }
+
+    /* Zero length thing name is not valid,
+     * $"$aws/things/<ThingName>/
+     * $"$aws/things/<ThingName>"
+     * will extract the same thing name result.
+     * Only empty thing name string like:
+     * "$aws/things/" or "$aws/things" will fail.
+     */
+    if( index > 0U )
+    {
+        * pThingName = pString;
+        * pThingNameLength = index;
+        returnStatus = SHADOW_STATUS_SUCCESS;
+    }
+
+    return returnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static ShadowStatus_t extractShadowMessageType( const char * pString,
+                                                uint16_t stringLength,
+                                                ShadowMessageType_t * pMessageType )
+{
+    uint32_t index = 0U;
+    ShadowStatus_t returnStatus = SHADOW_STATUS_FAIL;
+
+    /* Lookup table for Shadow message string. */
+    static const char * const pMessageStrings[ ShadowMessageTypeMaxNum ] =
+    {
+        SHADOW_TOPIC_OPERATION_STRING_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow message string length. */
+    static const uint16_t pMessageStringsLength[ ShadowMessageTypeMaxNum ] =
+    {
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow message types. */
+    static const ShadowMessageType_t pMessageTypes[ ShadowMessageTypeMaxNum ] =
+    {
+        ShadowMessageTypeGetAccepted ,
+        ShadowMessageTypeGetRejected,
+        ShadowMessageTypeDeleteAccepted,
+        ShadowMessageTypeDeleteRejected,
+        ShadowMessageTypeUpdateAccepted,
+        ShadowMessageTypeUpdateRejected,
+        ShadowMessageTypeUpdateDocument,
+        ShadowMessageTypeUpdateDelta
+    };
+
+    for( ; index < ( uint32_t ) ( sizeof( pMessageStrings ) / sizeof( pMessageStrings[0] ) ); index++ )
+    {
+        returnStatus = containsSubString( pString,
```

Comment:
Not sure if this is premature optimization, but I noticed all of these strings begin with "/shadow/". If we match the first 8 characters of `pString` to `"shadow/"` before entering the `for` loop, that should save us time during the `strncmp` that's in this function.

Reply:
I did consider to do the same thing while implementing it, but to jump to "/shadow/" and then do the compare, this will make us have to compare against a macro "/get/" or "/get/accepted" and others likewise.
Macros that doesn't have shadow, might need a new naming rule, and I'm afraid might cause some confusion.
Matching an incoming message is not very performance sensitive. I'm not sure it is worth it to this modification.

Please let me know if you still suggest to make this change.

Thank you.

Reply:
I agree that we can avoid adding more macros and checks here and live with it.